src/bootstrap/metrics.rs

```diff
@@ -0,0 +1,262 @@
+    }
+
+    pub(crate) fn persist(&self, build: &Build) {
+        let mut state = t!(self.state.lock());
+        assert!(state.running_steps.is_empty(), "steps are still executing");
+
+        let dest = build.out.join("build-metrics.json");
+
+        let mut system = System::new();
+        system.refresh_cpu();
+        system.refresh_memory();
+
+        let system_stats = JsonInvocationSystemStats {
+            cpu_threads_count: system.processors().len(),
+            cpu_model: system.processors()[0].brand().into(),
+
+            memory_total_bytes: system.total_memory() * 1024,
+        };
+        let steps = std::mem::take(&mut state.finished_steps);
+
+        // Some of our CI builds consist of multiple independent CI invocations. Ensure all the
+        // previous invocations are still present in the resulting file.
+        let mut invocations = match std::fs::read(&dest) {
+            Ok(contents) => t!(serde_json::from_slice::<JsonRoot>(&contents)).invocations,
+            Err(err) => {
+                if err.kind() != std::io::ErrorKind::NotFound {
+                    eprintln!("warning: failed to open existing metrics file: {err}");
+                }
+                Vec::new()
+            }
+        };
+        invocations.push(JsonInvocation {
+            duration_including_childs_sec: state.invocation_timer_start.elapsed().as_secs_f64(),
+            childs: steps.into_iter().map(|step| self.prepare_json_step(step)).collect(),
+        });
+
+        let json = JsonRoot { system_stats, invocations };
+
+        t!(std::fs::create_dir_all(dest.parent().unwrap()));
+        let mut file = BufWriter::new(t!(File::create(&dest)));
+        t!(serde_json::to_writer(&mut file, &json));
+    }
+
+    fn prepare_json_step(&self, step: StepMetrics) -> JsonNode {
+        JsonNode::RustbuildStep {
+            type_: step.type_,
+            debug_repr: step.debug_repr,
+
+            duration_excluding_childs_sec: step.duration_excluding_childs_sec.as_secs_f64(),
+            system_stats: JsonStepSystemStats {
+                cpu_sampling_freq: CPU_SAMPLING_INTERVAL.as_secs_f64(),
+                cpu_single_core: step.cpu_single_core,
+                cpu_all_cores: step.cpu_all_cores,
+            },
+
+            childs: step.childs.into_iter().map(|child| self.prepare_json_step(child)).collect(),
+        }
+    }
+}
+
+fn cpu_sampler_thread(state: Arc<Mutex<MetricsState>>, reset: Receiver<()>) {
+    let mut system = System::new();
+    system.refresh_cpu();
+
+    loop {
+        // Wait for one interval between samples, with the ability to reset the interval whenever a
+        // new step starts. This ensures the first sample for each step is collected one interval
+        // after the step started executing, instead of a random time between the start of the step
+        // and the expected interval (depending on when the last sample was collected, as part of
+        // the previous step).
```

Comment:
This comment sort of suggests that a step will always get a sample, but for short-lived steps that's not the case.

I think we also need to consider that at least in theory, we should be hitting the reset at step exit as well -- we probably often have situations where we have an outer step that maybe spends 1ms on CPU and an inner step spending 30s on CPU, for example, followed by processing by the outer step (say for 100ms). This sort of interleaving is why I'm a little doubtful about the reset trigger at all, since it doesn't really seem like it's going to produce good results to me.